bftengine/tests/bcstatetransfer/test_replica.hpp

```diff
@@ -52,6 +52,7 @@ class TestReplica : public IReplicaForStateTransfer {
 
   void changeStateTransferTimerPeriod(uint32_t timerPeriodMilli) override{};
 
+  concordUtil::Timers::Handle addOneShotTimer(uint32_t timeoutMilli) override { return concordUtil::Timers::Handle(); }
```

Comment:
whats is the purpose of the ```OneShotTimer``` ?

Reply:
The main reason is performance optimization. I'll explain via an example:  when putting block concurrently, let's say I have 4 jobs for putBlock running concurrently. None of them is ready yet. Let's assume that at moment the source is disconnected and no more ItemDataMsg arrives. Nothing will trigger the completion of these jobs, I still want to finalize these jobs ASAP. Without this timer, I will finalize them only on the next periodic timer, which is a long time from the current time.  With some logic (I am using a flag, and use this timer in rare events) I trigger a timer for 5 ms from the current time, to come back and check the future again. That way, all jobs will be finalized when switching to a different source or when (maybe) the current source is back (retransmission successful).

Again, this timer is not mandatory, in some rare cases, it will just improve performance.
Why not just wait on the jobs? since the main thread can do other things meanshile. I would like to add more optimisations such as digest computation in parallel etc. The main thread should always keep moving and only wait when it must wait, but keep moving to supervise other jobs and fetch messages from the incoming queue.